File: DSA_Exercise/Arrays/ThreeSum.java
// Give an array and a "target". You need to find all the triplets (a[i], a[j], a[k]) such that a[i]+ a[j]+ a[k] = target and i!=j!=k.
// Sample input 
// N=7
// arr = {7, -6, 3, 8, -1, 8, -1}
// target=0
// sample output = [-11,3,8],[-6,-1,7]
// Time complexity = O(n^2) 
// Space complexity = O(1)

package DSA_Exercise.Arrays;
import java.util.*;

public class ThreeSum{
    public static void main(String[] args){
        int[] arr = {7, -6, 3, 8, -1, 8, -11};
        int target = 0;
        solution(arr, target, arr.length);
    }
    public static void solution(int[] a, int target, int n){
        Arrays.sort(a);
        for(int i=0; i<n; i++){
            if(i==0 || a[i] != a[i-1]){
                int j=i+1, k=n-1;
                int tar = target-a[i];
                while (j<k){
                    if(a[j] + a[k] == tar){
                        System.out.println(a[i] + " " + a[j] + " " + a[k]);
                        while (j<k && a[j] == a[j+1]) j++;
                        while (j<k && a[k] == a[k-1]) k--;
                        j++;
                        k--;
                    }else if (a[j] + a[k] < tar){
                        j++;
                    }
                    else{
                        k--;
                    }
                }
            }
        }
    }
}

File: Array/ThreeSum.py
# Input: nums = [-1,0,1,2,-1,-4]
# Output: [[-1,-1,2],[-1,0,1]]
# Explanation: 
# nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
# nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
# nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
# The distinct triplets are [-1,0,1] and [-1,-1,2].
# Notice that the order of the output and the order of the triplets does not matter.
# Time Complexity: O(n)^2
# Space Complexity: 0(0)

from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # Sort the array
        nums.sort()
        result = []
        n = len(nums)
        
        # Iterate through the array
        for i in range(n):
            # Skip duplicate values for the fixed pointer
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            
            # Two-pointer search
            left, right = i + 1, n - 1
            while left < right:
                current_sum = nums[i] + nums[left] + nums[right]
                
                if current_sum == 0:
                    # Found a triplet
                    result.append([nums[i], nums[left], nums[right]])
                    
                    # Skip duplicates for the left pointer
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    # Skip duplicates for the right pointer
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    
                    # Move both pointers
                    left += 1
                    right -= 1
                elif current_sum < 0:
                    # Increase the sum by moving the left pointer
                    left += 1
                else:
                    # Decrease the sum by moving the right pointer
                    right -= 1
        
        return result